### PR TITLE
Add screenshot api

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Fixes https://github.com/bevyengine/bevy/issues/1207

# Objective

Right now, it's impossible to capture a screenshot of the entire window without forking bevy. This is because
- The swapchain texture never has the COPY_SRC usage
- It can't be accessed without taking ownership of it
- Taking ownership of it breaks *a lot* of stuff

## Solution

- Introduce a dedicated api for taking a screenshot of a given bevy window, and guarantee this screenshot will always match up with what gets put on the screen.

---

## Changelog

- Added the `ScreenshotManager` resource with two functions, `take_screenshot` and `save_screenshot_to_disk`

### CODE REVIEW

File: crates/bevy_render/src/view/window.rs
Comment (StarArawn) [inline crates/bevy_render/src/view/window.rs:141]: Probably should use `try_lock` here to avoid a deadlock if a screenshot is being taken at the same time? Although I guess the bevy scheduler prevents that from happening here..

**Review by StarArawn — CHANGES_REQUESTED** [review]
Looks pretty good over all. Can we add an example showcasing this? Right now it's a little hard to test and verify the work.

File: crates/bevy_render/src/lib.rs
Comment (IceSentry) [inline crates/bevy_render/src/lib.rs:140]: Should this be behind a feature flag? I assume it's not a big deal, but it doesn't really need to be present all the time either.

Comment (IceSentry) [reply crates/bevy_render/src/lib.rs:140]: Actually, I'm curious if the entire feature could be behind a flag? I don't want to block on that, I'm just curious if it's possible/if it's even worth it.

Comment (james7132) [reply crates/bevy_render/src/lib.rs:140]: Seems generally useful for rendering testing as it would allow for golden image testing, and it seems fairly trivial for a user to make a player facing screenshot keyboard shortcut too. Unless I'm seeing a hidden cost here. seems like a general win to me.

Comment (IceSentry) [reply crates/bevy_render/src/lib.rs:140]: Yeah, I was more thinking about not needing to compile it if it's unused, but this is probably too small to be worth the complexity of a feature flag.

Comment (IceSentry) [reply crates/bevy_render/src/lib.rs:140]: I didn't think about a player screenshot feature though, good idea.

Comment (targrub) [reply crates/bevy_render/src/lib.rs:140]: A cost is the new dependency on `bevy_task`.

File: crates/bevy_render/src/view/window.rs
Comment (TheRawMeatball) [reply crates/bevy_render/src/view/window.rs:141]: Yep, this can't really happen because `callbacks` is `pub(crate)` and this is the singular callsite where it's called. Even if somehow a user ended up with two of these systems, one would simply take the lock and they'd run in series. Only one lock isn't enough to cause a deadlock.

File: crates/bevy_render/src/lib.rs
Comment (TheRawMeatball) [reply crates/bevy_render/src/lib.rs:140]: bevy_tasks was already a dependency through bevy_ecs, this doesn't really change anything.

File: crates/bevy_render/src/texture/image_texture_conversion.rs
Comment (DevinLeamy) [inline crates/bevy_render/src/texture/image_texture_conversion.rs:200]: Why do we introduce the `TextureFormat::Bgra8UnormSrgb` here? Is there a reason `bgra` might be preferred over `rgba`?

File: crates/bevy_render/src/lib.rs
Comment (alice-i-cecile) [reply crates/bevy_render/src/lib.rs:140]: Yeah I don't feel this needs to be feature flagged.

File: crates/bevy_render/src/texture/image_texture_conversion.rs
Comment (alice-i-cecile) [reply crates/bevy_render/src/texture/image_texture_conversion.rs:200]: Seconding the confusion about why this is part of this PR. I don't mind it, just a bit confused.

File: crates/bevy_render/src/view/window.rs
Comment (alice-i-cecile) [reply crates/bevy_render/src/view/window.rs:141]: Nile, your comment here would make a great comment for this line of code.

File: crates/bevy_render/src/texture/image_texture_conversion.rs
Comment (TheRawMeatball) [reply crates/bevy_render/src/texture/image_texture_conversion.rs:200]: wgpu apparently quite likes this format for swapchain textures - this was necessary to get the save to disk functionality working on my machine.

Comment (alice-i-cecile) [reply crates/bevy_render/src/texture/image_texture_conversion.rs:200]: Interesting. Can you please leave a comment here to that effect then?

**Comment by mockersf** [comment]
crashes for me on macOS:
```
thread 'main' panicked at 'Error in Surface::configure: requested usage is not supported', wgpu-0.14.2/src/backend/direct.rs:274:9
```

**Comment by TheRawMeatball** [comment]
Changed the code to ensure a panic like the one @mockersf encountered is no longer possible, but it's a lot more code now.

File: crates/bevy_render/src/view/window.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/window.rs:311]: I'd recommend adding labels here and the rest of the code. It makes things way easier to debug in renderdoc when everything is labelled.

File: crates/bevy_render/src/view/window.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/window.rs:276]: Not super important, but this should probably be a method on window since it's repeated 3 times.

File: crates/bevy_render/src/view/window/screenshot.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/window/screenshot.rs:62]: would it be possible to bubbleup the error or at least use expect here? I feel like hitting a generic image conversion error might be confusing if someone is just trying to take a screenshot and it fails here.

File: crates/bevy_render/src/view/window/screenshot.wgsl
Comment (IceSentry) [inline crates/bevy_render/src/view/window/screenshot.wgsl:2]: I'm not sure exactly what this is doing, could you maybe add a comment explaining it?

**Comment by mockersf** [comment]
> Changed the code to ensure a panic like the one @mockersf encountered is no longer possible, but it's a lot more code now.

Works for me now! Thanks

**Comment by mockersf** [comment]
There is an issue around screenshooting something with bloom enabled. I enabled bloom and HDR in example `animated_fox`

## OS Screenshot

Screenshot 2023-01-12 at 22 47 47

## This PR Screenshot

test

File: crates/bevy_render/src/view/window/screenshot.rs
Comment (mockersf) [inline crates/bevy_render/src/view/window/screenshot.rs:146]:
```suggestion
                    format: TextureFormat::bevy_default(),
```
It's wanting a `Bgra8UnormSrgb` because that's what defined by the `bevy_default` method

**Comment by mockersf** [comment]
Crashes in WASM too:
```
Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 726, Gl)>`
    In a set_pipeline command
      note: render pipeline = `<RenderPipeline-(0, 1, Gl)>`
    Render pipeline targets are incompatible with render pass
    Incompatible color attachment: the renderpass expected [Some(Rgba8UnormSrgb)] but was given [Some(Bgra8UnormSrgb)]
```

File: crates/bevy_render/src/renderer/mod.rs
Comment (TheRawMeatball) [reply crates/bevy_render/src/renderer/mod.rs:37]: Not really, other than not wanting to pollute the namespace i guess.

**Comment by TheRawMeatball** [comment]
> There is an issue around screenshooting something with HDR & bloom enabled. I enabled them in example `animated_fox`.

This seems to be a result of the brightness values being stored in the alpha channel - the two images are once again identical when alpha is disabled.

**Comment by TheRawMeatball** [comment]
Only fixed this fix for the save_screenshot_to_disk function, as being able to tap into this data could be useful.

**Comment by TheRawMeatball** [comment]
@mockersf I'd appreciate if you could test wasm and especially hdr again, I don't have any hdr hardware so I can't be 100% sure stuff is working.

File: crates/bevy_render/src/view/window/screenshot.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/window/screenshot.rs:163]: since you are using a fullscreen triangle, I'd suggest using `bevy_core_pipeline::fullscreen_vertex_shader::fullscreen_shader_vertex_state() ` instead of doing it manually. That way you can also remove the vertex shader from your wgsl.

File: crates/bevy_render/src/renderer/mod.rs
Comment (IceSentry) [reply crates/bevy_render/src/renderer/mod.rs:37]: Fair, I don't mind either way. I'd prefer having the `screenshot` module in scope just to not have `crate::view`, but really it's up to you, really not a big deal.

File: crates/bevy_render/src/view/window/screenshot.rs
Comment (TheRawMeatball) [reply crates/bevy_render/src/view/window/screenshot.rs:163]: Can't - this code needs to live in bevy_render, which can't have a dependency on bevy_core_pipeline.

Comment (IceSentry) [reply crates/bevy_render/src/view/window/screenshot.rs:163]: Oh right, didn't think about that.

File: crates/bevy_render/src/view/window/screenshot.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/window/screenshot.rs:250]: Can you explain a little bit what this is doing? I'm not entirely sure I understand what's going on here.

Comment (TheRawMeatball) [reply crates/bevy_render/src/view/window/screenshot.rs:250]: This function basically loops over all the windows, sees which ones have had a screenshot taken, and creates async tasks per window with a screenshot to copy that memory from the GPU to the CPU, and eventually calls the relevant screenshot callback.

In a way it's "collecting" the screenshots that were taken to the CPU to continue processing

**Comment by superdump** [comment]
This looks like it should fix #22

**Comment by superdump** [comment]
And #1207

**Comment by alice-i-cecile** [comment]
bors try

**Comment by alice-i-cecile** [comment]
@TheRawMeatball can you note that this will fix the issues mentioned above?

**Comment by TheRawMeatball** [comment]
I didn't add or remove any dependencies, the dependency failiure shouldn't be from this PR, and yes, all issues mentioned above should be solved.

**Comment by aevyrie** [comment]
CI is failing on https://github.com/bevyengine/bevy/issues/7654

**Comment by alice-i-cecile** [comment]
> Thanks for putting this together, @TheRawMeatball. What are the blockers for merging this?

This needs review mostly, especially from folks with rendering expertise. This ended up trickier than I would have expected.

Community review counts, so please feel free to go over this carefully and approve it if you think it's ready :)

**Comment by mockersf** [comment]
the PR also needs to be updated for #7681 I think

**Comment by hymm** [comment]
What's the current status of this PR?

**Comment by WannaBeKratos** [comment]
Does this snippet work without the WinitPlugin ?
 `        // camera
        commands.spawn(Camera3dBundle {
            camera_3d: Camera3d {
                clear_color: ClearColorConfig::Custom(Color::WHITE),
                ..default()
            },
            camera: Camera {
                // render before the "main pass" camera
                order: -1,
                target: RenderTarget::Image(image_handle),
                ..default()
            },
            transform: Transform::from_xyz(0.0, 50.0, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
            ..default()
        });
    }`

**Comment by mockersf** [comment]
@TheRawMeatball I think there has been an issue with your merge

**Comment by DGriffin91** [comment]
What's the purpose of the render pass? Is it to avoid putting `TextureUsages::COPY_SRC` on the render target?

File: crates/bevy_render/src/view/window/screenshot.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/window/screenshot.rs:74]: There should probably be an info log here with the path to let people know a screenshot was taken.

**Review by IceSentry — APPROVED** [review]
I feel like there should probably be a bit more docs in screenshot.rs, but in general the code works and I haven't seen anything obviously wrong.

I don't fully like the idea of passing a finalizer callback to the graph runner, but I'm not sure how to avoid that.

**Review by mockersf — APPROVED** [review]

**Review by cart — APPROVED** [review]
I think this is a reasonable solution to the specific problem of taking screenshots of a window. But I don't love how "hacked in" it is. I think this impl points to a fundamental "render output feedback" problem that should (ultimately) be solved generically. The screenshot api is pretty opinionated and isn't a full solve to the problem. People should be able to build similar abstractions on their own.

I see a number of paths forward:
* Continue down the `finalizer` path, but just modularize it (ex let people add arbitrary functions to the process via some new abstraction) 
* Build this into our graph abstraction ... this would mean screenshotting would not be a global feature, but rather one that exists in specific graphs ... this seems ok to me. We could build common "feedback" nodes (including a screenshot node). It seems to me like this PR could have already been expressed that way, rather than adding the `finalizer`. Functionally `submit_screenshot_commands` is really equivalent to a "render graph node that exists in every graph and always runs last". This doesn't solve the "window / swap chain / image management stuff" though, which is definitely the bigger challenge. It feels like a chunk of this could/should be expressed as some generic graph node that says "at this point in the graph, write this COPY_SRC texture view to a buffer of appropriate size and (maybe optionally) create an image from it". That leaves the problem of "writing the swap chain to a COPY_SRC texture" to be solved on its own in whatever way we deem "best" / "most modular"/ whatever we choose to optimize for.
* Give up on reading swap chain outputs and instead read `ViewTarget::maint_target_textures` outputs in render graph nodes. Seems like we could make those COPY_SRC (maybe in an opt-in fashion when screenshots are enabled). This does ignore the upscaling step though (which does seem like a problem).

That being said, creating a generic abstraction is not a blocker (and I understand there is a fair amount of implementation nuance here when it comes to swapchain access). This is totally fine for now.

**Comment by zimond** [comment]
Still could this work without a window? It seems some questions in the comments are unanswered.

**Pull request merged**
